### PR TITLE
feat: client pagination enhancement

### PR DESCRIPTION
**Pull request by pujavs**
Hello @yuriyz, the change is to enable pagination for openid client request.

Client ticket ref: https://support.gluu.org/other/10067/clients-paging/

### CODE REVIEW

File: common/src/main/java/io/jans/configapi/util/ApiConstants.java [hunk 1]
```diff
@@ -100,5 +101,12 @@
     public static final int CONNECTION_POOL_DEFAULT_MAX_PER_ROUTE = 20;
     public static final int CONNECTION_POOL_VALIDATE_AFTER_INACTIVITY = 10;
     public static final int CONNECTION_POOL_CUSTOM_KEEP_ALIVE_TIMEOUT = 5;
+    
+    //Pagination
+    public static final String DEFAULT_LIST_SIZE = "50";
+    public static final String DEFAULT_LIST_START_INDEX = "1";
```
Comment (yuriyz) [inline common/src/main/java/io/jans/configapi/util/ApiConstants.java:107]: Size and index has to be `int`, not `String`.

File: server/src/main/java/io/jans/configapi/rest/resource/auth/BaseResource.java [hunk 2]
```diff
@@ -65,4 +81,57 @@ protected static Response getNotAcceptableException(String msg) {
...
 
+    protected Response prepareSearchRequest(String schemas, String filter, String sortBy, String sortOrder,
+            Integer startIndex, Integer count, String attrsList, String excludedAttrsList, SearchRequest request) {
+        log.debug(
+                "Search Request params:: - schemas:{}, filter:{}, sortBy:{}, sortOrder:{}, startIndex:{}, count:{}, attrsList:{}, excludedAttrsList:{}, request:{} ",
+                schemas, filter, sortBy, sortOrder, startIndex, count, attrsList, excludedAttrsList, request);
+
+        Response response = null;
+
+        if (StringUtils.isNotEmpty(schemas)) {
```
Comment (yuriyz) [inline server/src/main/java/io/jans/configapi/rest/resource/auth/BaseResource.java:92]: As in previos PRs, when we make "early exit" we don't have all these block nestings.
```
if (StringUtils.isEmpty(schemas)) {
    return Response.status(Response.Status.BAD_REQUEST).entity("Schema(s) not supplied in Search Request")
                    .build();    
}
```

File: server/src/main/java/io/jans/configapi/rest/resource/auth/BaseResource.java [hunk 3]
```diff
@@ -65,4 +81,57 @@ protected static Response getNotAcceptableException(String msg) {
...
+            count = count == null ? getMaxCount() : count;
+            log.debug(" count:{} ", count);
+            // Per spec, a negative value SHALL be interpreted as "0" for count
+            if (count < 0) {
+                count = 0;
+            }
+
+            if (count <= getMaxCount()) {
```
Comment (yuriyz) [inline server/src/main/java/io/jans/configapi/rest/resource/auth/BaseResource.java:100]:
```
if (count > getMaxCountLimit())  {
     return Response.status(Response.Status.BAD_REQUEST)
                        .entity("Maximum number of results per page is " + getMaxCount()).build();
}

```

File: server/src/main/java/io/jans/configapi/rest/resource/auth/ClientsResource.java [hunk 4]
```diff
@@ -58,20 +60,29 @@
...
-            clients = clientService.getAllClients(limit);
-        }
+            @DefaultValue("") @QueryParam(value = ApiConstants.PATTERN) String pattern,
+            @DefaultValue(DEFAULT_LIST_START_INDEX) @QueryParam(value = ApiConstants.START_INDEX) int startIndex,
+            @QueryParam(value = ApiConstants.SORT_BY) String sortBy,
+            @QueryParam(value = ApiConstants.SORT_ORDER) String sortOrder) throws EncryptionException {
+        logger.debug("Client serach param - limit:{}, pattern:{}, startIndex:{}, sortBy:{}, sortOrder:{}", limit, pattern,
+                startIndex, sortBy, sortOrder);
+
+        SearchRequest searchReq = new SearchRequest();
```
Comment (yuriyz) [inline server/src/main/java/io/jans/configapi/rest/resource/auth/ClientsResource.java:70]: `SearchRequest searchReq = createSearchRequest(...)`

**Review by yuriyz — CHANGES_REQUESTED** [review]
See inlined comments.

File: common/src/main/java/io/jans/configapi/model/configuration/ApiAppConfiguration.java [hunk 5]
```diff
@@ -31,6 +31,7 @@
     private String loggingLayout;
     private String externalLoggerConfiguration;
     private Boolean disableJdkLogger = true;
+    private int maxCount;
```
[the remark this answers is not in the document]
Comment (pujavs) [reply common/src/main/java/io/jans/configapi/model/configuration/ApiAppConfiguration.java:34]: It is configuration max count which will be read from DB.

File: common/src/main/java/io/jans/configapi/util/ApiConstants.java (hunk 1 above)
Comment (pujavs) [reply common/src/main/java/io/jans/configapi/util/ApiConstants.java:107]: These are used a javax.ws.rs.DefaultValue which takes value as String.

File: server/src/main/java/io/jans/configapi/rest/resource/auth/BaseResource.java [hunk 6]
```diff
@@ -65,4 +81,57 @@ protected static Response getNotAcceptableException(String msg) {
...
+
+                request.setSchemas(schemas);
+                request.setAttributes(attrsList);
+                request.setExcludedAttributes(excludedAttrsList);
+                request.setFilter(filter);
+                request.setSortBy(sortBy);
+                request.setSortOrder(sortOrder);
+                request.setStartIndex(startIndex);
+                request.setCount(count);
+                request.setMaxCount(getMaxCount());
```
[the remark this answers is not in the document]
Comment (pujavs) [reply server/src/main/java/io/jans/configapi/rest/resource/auth/BaseResource.java:116]: request variable is passed as reference value to the methods and hence the values set inside this method is available to the caller

File: server/src/main/java/io/jans/configapi/rest/resource/auth/BaseResource.java [hunk 7]
```diff
@@ -65,4 +81,57 @@ protected static Response getNotAcceptableException(String msg) {
         return Response.status(Response.Status.NOT_ACCEPTABLE).entity(error).build();
     }
 
+    protected Response prepareSearchRequest(String schemas, String filter, String sortBy, String sortOrder,
```
[the remark this answers is not in the document]
Comment (pujavs) [reply server/src/main/java/io/jans/configapi/rest/resource/auth/BaseResource.java:84]: Refactored the method

File: server/src/main/java/io/jans/configapi/rest/resource/auth/BaseResource.java (hunk 3 above)
Comment (pujavs) [reply server/src/main/java/io/jans/configapi/rest/resource/auth/BaseResource.java:100]: Done

File: server/src/main/java/io/jans/configapi/rest/resource/auth/ClientsResource.java (hunk 4 above)
Comment (pujavs) [reply server/src/main/java/io/jans/configapi/rest/resource/auth/ClientsResource.java:70]: Was not possible with original code as prepareSearchRequest was returning Response but now with refactored code changing as suggested

File: server/src/main/java/io/jans/configapi/rest/resource/auth/BaseResource.java (hunk 2 above)
Comment (pujavs) [reply server/src/main/java/io/jans/configapi/rest/resource/auth/BaseResource.java:92]: Done

File: server/src/main/java/io/jans/configapi/rest/resource/auth/BaseResource.java [hunk 8]
```diff
@@ -65,4 +81,57 @@ protected static Response getNotAcceptableException(String msg) {
...
+            Integer startIndex, Integer count, String attrsList, String excludedAttrsList, SearchRequest request) {
+        log.debug(
+                "Search Request params:: - schemas:{}, filter:{}, sortBy:{}, sortOrder:{}, startIndex:{}, count:{}, attrsList:{}, excludedAttrsList:{}, request:{} ",
+                schemas, filter, sortBy, sortOrder, startIndex, count, attrsList, excludedAttrsList, request);
+
+        Response response = null;
```
[the remark this answers is not in the document]
Comment (pujavs) [reply server/src/main/java/io/jans/configapi/rest/resource/auth/BaseResource.java:90]: not required, removed

File: server/src/main/java/io/jans/configapi/rest/resource/auth/ClientsResource.java [hunk 9]
```diff
@@ -167,4 +178,20 @@ private String generatePassword() throws NoSuchAlgorithmException {
...
     }
+
+    private List<Client> doSearch(SearchRequest searchReq) {
+        logger.debug("Client search params - searchReq:{} ", searchReq);
+
+        PagedResult<Client> pagedResult = clientService.searchClients(searchReq);
+        log.trace("PagedResult  - pagedResult:{}", pagedResult);
+
+        List<Client> clients = null;
+        if (pagedResult != null) {
```
[the remark this answers is not in the document]
Comment (pujavs) [reply server/src/main/java/io/jans/configapi/rest/resource/auth/ClientsResource.java:189]: Done

**Review by yuriyz — APPROVED** [review]

**Pull request merged**